---
fontsize: 12pt
geometry: margin=0.50in
subparagraph: yes
params:
    title: "title"
    date: "0000-00-00"
title: "`r params$title`"
date:  "`r params$date`"
header-includes:
   - \usepackage{makecell}
   - \usepackage{booktabs}
   - \usepackage{amsmath}
output: 
  pdf_document:
    fig_caption: yes
    toc: true
    keep_tex: no
    latex_engine: xelatex
---

```{r loading, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
library(RColorBrewer)
library(dplyr)
library(kableExtra)
library(argparse)
library(GenomicRanges)

options(tinytex.tlmgr_update = FALSE)
knitr::opts_chunk$set(echo = TRUE)
```

\newpage

# Introduction

The attached report describes the results of integration site analysis for samples from gene therapy trials. Cellular DNA was processed as described in Sherman et al. 2017 (doi: 10.1016/j.omtm.2016.11.002.) and analyzed using the analysis software AAVengeR (https://github.com/helixscript/AAVengeR). 

# PI Summary

```{r piSummary, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}

if (!is.null(args$piNote)) {
  file_content <- readLines(args$piNote)
  cat(file_content, sep = "\n")
} else {
  cat("No Summary Given")
}

```

\newpage

# Sequencing and Integration Summary Table

Presented below are summary details of each sample. We estimate the number of cell clones sampled using the SonicLength method (Berry, 2012); this is summarized in the column “Inferred cells”. Relative abundance was not measured from read counts, which are known to be inaccurate, but
from marks introduced into DNA specimens prior to PCR amplification using the SonicLength method
PMID:22238265.

Under most circumstances, only a subset of sites will be sampled. We thus include an estimate of sample size based on the frequency of isolation information from the SonicLength method (Berry, 2012). The ‘S.chao1’ column denotes the estimated lower bound for population size derived using Chao estimate (Chao, 1987).

```{r summaryTable, echo=FALSE}
if(nrow(summary) > 0) {
  kable(summary, "latex",  linesep = "", booktabs = T, caption = "integration summary table") %>%
    kable_styling(latex_options = c("hold_position", "scale_down"))
} else {
  print("error making summary table")
}
```

\newpage

# Abundance Frequency Plots

Stacked bar plots indicate clonal abundance frequency. Only the top 10 clonal types are being plotted, and the rest will be ploted in grey as "low abundance". The number above the stacked bar plots indicates the total clonal types. The legend in each plot correspond to the clonal type integration sites. The number after colon indicates the count for that specific clonal type.

```{r abundantPlot, echo=FALSE, message=FALSE, warning=FALSE,fig.height=4.5, fig.width=4}


for (i in abundantPlot) {
  print(i)
}

```

\newpage

# AAV ITR Breakpoint Plots

These plots are a visual representation of the ITR breakpoint at the site of genomic integration. The arrow indicates the priming location within the ITR and the x-coordinate of each bar corresponds to a departure from an expected ITR sequence. Each departure from expectation is demarcated as a rearrangement, and bars are colored to reflect the frequency of rearrangement at each integration site. 

```{r itrPlots, echo=FALSE, message=FALSE, warning=FALSE, fig.height=4, fig.align='center',results='asis', fig.width=9.5, dev='cairo_pdf'}
 
for (i in remnant.plot){ 
  print(i)
}

```

\newpage

# Gene distribution dot plots

The plot below details the % of integration sites within two features, 'in genes' and 'in Exons', summarized per sample. 'in Genes' refers to sites within transcription unit boundaries and 'in Exons' refers to sites within exon boundaries. 

```{r geneTable, echo=FALSE}
if(nrow(gene.dist) > 0) {
  tmp <- gene.dist %>% 
    select(-description) %>%
    relocate(info, .after = sample)
  
  kable(tmp, "latex", linesep = "", booktabs = T, align = "c", 
        caption = "gene distribution percentage table") %>%
  kable_styling(latex_options = "hold_position")
} else {
  print("error making gene distribution table")
}
```

```{r genePlot, echo=FALSE, fig.width=9, fig.height=5, fig.cap="gene distribution scatter plot"}

ggplot(gene.dist, aes(y = description)) +
  geom_point(aes(x = `In Exon%`, shape = "exon"), alpha = 0.8) +    
  geom_point(aes(x = `In Transcription Unit%`, shape = "gene"), alpha = 0.8, size = 2) + 
  scale_shape_manual(values = c(16, 2), labels = c("exon", "TranscriptionUnit"), name = NULL) + 
  labs(title = "", x = "") + 
  scale_x_continuous(labels = scales::label_percent(accuracy = .1, scale = 1), 
                     limits = c(0, 100)) +
  geom_vline(linetype = "dashed", aes(xintercept = dash[[1]], color = "random gene")) +
  geom_vline(linetype = "dotted", aes(xintercept = dash[[2]], color = "random TU")) +
  theme(axis.text = element_text(size = 10),                  
        axis.title.y = element_blank(),
        panel.grid.major = element_blank(),                   
        axis.line = element_line(linewidth = 0.5),
        legend.position = "right")

```

\clearpage
\newpage

# Methods

Report Generated on: `r params$date` \newline AAVenger Version: 1.1 \newline 
\newline 
Modules Called: 

* core.R  
* mapSiteLeaderSequences.R  
* buildAAVremnantPlots.R  
* anchorReadRearrangements.R


