---
fontsize: 12pt
geometry: margin=0.50in
subparagraph: yes
params:
    title: "title"
    date: "0000-00-00"
title: "`r params$title`"
date:  "`r params$date`"
header-includes:
   - \usepackage{makecell}
   - \usepackage{booktabs}
   - \usepackage{amsmath}
output: 
  pdf_document:
    fig_caption: yes
    toc: true
    keep_tex: no
    latex_engine: xelatex
---

```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
library(RColorBrewer)
library(dplyr)
library(kableExtra)

#library(vegan)
#library(xtable)
#library(gridExtra)
#library(grDevices)
#library(grid)
options(stringsAsFactors = FALSE)
options(tinytex.tlmgr_update = FALSE)
knitr::opts_chunk$set(echo = TRUE)
```

\newpage

# Introduction

The attached report describes results of analysis of integration site distributions and relative abundance for samples from gene therapy trials.

# Sample Summary

pull info from gtsp database for summary table.

# Sequencing and Integration Summary Table

Presented below are summary details of each sample. The Unique Molecular Identifiers (UMIs) estimate the number of unique molecules queried. Unique Sites refers to the number of unique integration sites detected. Chao1 estimates the total population sizes of integration sites in the sample.

Prelimary work with negative controls support the use of a filtering parameter in such that sites only supported by one read are considered artifactual and removed.

```{r summaryTable, echo=FALSE}
if(nrow(summary) > 0) {
  kable(summary, "latex",  linesep = "", booktabs = T, caption = "integration summary table") %>%
    kable_styling(latex_options = "hold_position")
} else {
  print("error making summary table")
}
```

\newpage

# Abundance Frequency Plots

Abundance 

```{r abundantPlot, echo=FALSE, message=FALSE, warning=FALSE,fig.height=4.5, fig.width=4}


for (i in abundantPlot) {
  print(i)
}

```

\newpage

# AAV ITR Breakpoint Plots

These plots are a visual representation of the ITR integration breakpoint. The x-axis indicates the point of departure from the expected ITR, where zero refers to the annotated start of the D-region. The arrow indicates the location within the ITR of the sequencing primer and the dotted lines indicate the tips of the B and C dumbells. Colors indicate the number of rearrangements in the ITR remnant.

```{r itrPlots, echo=FALSE, message=FALSE, warning=FALSE, fig.height=4, fig.align='center',results='asis', fig.width=10}
 
for (i in remnant.plot){ 
  print(i)
}

```

\newpage

# Gene distribution dot plots

Gene distribution for integration sites.

```{r geneTable, echo=FALSE}
if(nrow(gene.dist) > 0) {
  kable(gene.dist, "latex", linesep = "", booktabs = T, align = "c", 
        caption = "gene distribution percentage table") %>%
  kable_styling(latex_options = "hold_position")
} else {
  print("error making gene distribution table")
}
```


```{r genePlot, echo=FALSE, fig.width=9, fig.cap="gene distribution scatter plot"}

ggplot(gene.dist, aes(y = sample)) +
  geom_point(aes(x = `In Exon %`, shape = "exon"), alpha = 0.8) +    
  geom_point(aes(x = `In Gene %`, shape = "gene"), alpha = 0.8, size = 2) + 
  scale_shape_manual(values = c(16, 2), labels = c("exon", "gene"), name = NULL) + # Label legend, Remove legend title
  labs(title = "", x = "") + 
  scale_x_continuous(labels = scales::percent_format(scale = 1)) +
  theme(#plot.title = element_text(size = 16, face = "bold", hjust = 0.5),
        axis.text = element_text(size = 10),                  
        axis.title.y = element_blank(),
        panel.grid.major = element_blank(),                   
        axis.line = element_line(linewidth = 0.5),
        legend.position = "right")

```

\newpage

# Methods

Report Generated on: `r params$date` \newline AAVenger Version: 1.1 \newline 
\newline 
Modules Called: 

* core.R  
* mapSiteLeaderSequences.R  
* buildAAVremnantPlots.R  
* anchorReadRearrangements.R


