---
fontsize: 12pt
geometry: margin=0.50in
subparagraph: yes
params:
    title: "title"
    date: "0000-00-00"
    chromosome: ""
    chrom_sizes: ""
    sites: ""
    chaoTable: ""
    cutout: TRUE
title: "`r params$title`"
date:  "`r params$date`"
output: 
  pdf_document:
    toc: true
    keep_tex: no
    latex_engine: xelatex
---

```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(vegan)
library(knitr)
library(xtable)
library(ggforce)
library(gridExtra)
library(RColorBrewer)
library(grDevices)
library(wordcloud)
library(png)
library(grid)
library(pander)
library(tidyverse)
library(kableExtra)
options(stringsAsFactors = FALSE)
library(tiff)
library(grid)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

\newpage

# Introduction

There has been interest in the relationship between second-strand synthesis of AAV genomes and the DNA repair pathways non-homologous end joining (NHEJ) and homology-directed repair (HDR). These pathways may also play a role in the integrity of ITR-genome junctions in integrated vectors.

This preliminary study aims to understand the relationship between HDR and AAV integration integrity. U2OS cells were treated with DMSO or HDR inhibitor B02 (RAD51) prior to AAV transduction. The contribution of purification of the AAV preparations was also explored.

Samples are classified as follows:\newline DMSOCrude: Cells treated with DMSO and transduced with a crude AAV prep (GTSP5787, GTSP5792) \newline DMSOPure: Cells treated with DMSO and transduced with a pure AAV prep (GTSP5790, GTSP5795)\newline HDRiCrude: Cells treated with B02 and transduced with a crude AAV prep (GTSP5786, GTSP5791)\newline HDRiPure: Cells treated with B02 and transduced with a pure AAV prep(GTSP5789, GTSP5794)\newline Negative: Negative Control

Samples were also run with a positive control containing 6 synthetic yeast integration sites, each with breakpoints at the various regions of the ITR dumbbell (DA, DAB, DABB', DABB'C', DABB'C'C, DABB'C'CA)


## Data Generation

DNA extracted from cells were sequenced using a primer landing in the DA region of the AAV ITR. After alignment to the vector file, reads containing at least one break in the expected ITR sequence where the sequence following the break matched the vector sequence were selected. 

Note: To facilitate alignment, in the vector file provided LacO repeats were condensed to one of each repeat type.


\newpage

# Total Rearrangements

This output takes all the ITR containing reads and evaluates them for alignment to the vector. Any deviation from the expected ITR sequence is considered a rearrangement. The plot below depicts the percentage of unique molecules with an ITR remnant containing at least one rearrangement in the various Drug+AAV Prep Groups

```{r, echo=FALSE, message=FALSE, warning=FALSE}
knitr::include_graphics("DataPNG/Overall Rearrangement.png")
```

\newpage

# AAV ITR Breakpoint Plots

These plots are a visual representation of the the ITR structure and break points. The x-axis indicates the point of departure from the expected ITR, where zero refers to the annotated start of the D-region and the arrow indicates the location within the ITR of the sequencing primer. The height of the bar represents the number of unique vector positions post transition and the color of each bar represents the number of deviations from expected structure following the transition into vector. The % above each bar represents the % of breaks that occur in that segment of the ITR over all segments of the ITR.

In these diagrams, the flip and flop forms are condensed into a primary and secondary dumbbell. T The red region represents the DA region. The blue regions represent the primary dumbbell following the DA region. The green regions represent the secondary dumbbell and subsequent A' region. The dotted lines indicate the tips of the first and second dumbbells.

The sequence listed represents the sequence of the Flip 5' ITR form.

\newpage

## DMSOCrude

```{r echo=FALSE, out.width = "100%"}
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5787_itr.png")
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5792_itr.png")
```

\newpage

## DMSOPure

```{r echo=FALSE, out.width = "100%"}
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5790_itr.png")
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5795_itr.png")
```

\newpage

## HDRiCrude

```{r echo=FALSE, out.width = "100%"}
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5786_itr.png")
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5791_itr.png")
```

\newpage

## HDRiPure

```{r echo=FALSE, out.width = "100%"}
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5789_itr.png")
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5794_itr.png")
```

\newpage

# ITR Breakpoint Summary Graphs

## % By Region
This output takes all the percentages of each breakpoint and summarizes them by region. Samples are organized by Drug Condition, then AAV prep type.

```{r, echo=FALSE, message=FALSE, warning=FALSE,out.width = "90%"}
knitr::include_graphics("DataPNG/Breakdown of Each Region.png")
```

## % By Drug Condition
This output takes all the percentages of each break point and summarizes them by the % of breaks before the first ITR dumbbell. Samples are organized by Drug Condition and AAV prep type is not differentiated. Unpaired t-test was used to determined statistical significance (n=4), though significance should be considered primarily for hypothesis generation due to prep variability. 

```{r, echo=FALSE, message=FALSE, warning=FALSE,out.width = "50%"}
knitr::include_graphics("DataPNG/Statistics Across All Samples.png")
```
\newpage

# AAV Vector Breakpoint Plots

These plots are a visual representation of the structure after the ITR break point. The x-axis indicates the point in the vector where the read transitions out of the ITR into the vector. The height of the bar represents the number of reads that break into the vector position post transition and the color of each bar represents the number of deviations from expected structure following the transition into vector. 

In these diagrams, the flip and flop forms are both considered in the ITR region.The  blue LacO region (123 nucleotides) represents the 2371 nucleotides of repeats in the original vector. 


\newpage

## DMSOCrude

```{r echo=FALSE, out.width = "100%"}
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5787_vec.png")
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5792_vec.png")
```

\newpage

## DMSOPure

```{r echo=FALSE, out.width = "100%"}
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5790_vec.png")
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5795_vec.png")
```

\newpage

## HDRiCrude

```{r echo=FALSE, out.width = "100%"}
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5786_vec.png")
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5791_vec.png")
```

\newpage

## HDRiPure

```{r echo=FALSE, out.width = "100%"}
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5789_vec.png")
knitr::include_graphics("DataPNG/Annotated Breakpoint Maps_5794_vec.png")
```

\newpage

\newpage



# 
