---
fontsize: 12pt
geometry: margin=0.50in
subparagraph: yes
params:
    title: "title"
    date: "0000-00-00"
    chromosome: ""
    chrom_sizes: ""
    sites: ""
    chaoTable: ""
    cutout: TRUE
title: "`r params$title`"
date:  "`r params$date`"
output: 
  pdf_document:
    toc: true
    keep_tex: no
    latex_engine: xelatex
---

```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(vegan)
library(knitr)
library(xtable)
library(ggforce)
library(gridExtra)
library(RColorBrewer)
library(grDevices)
library(wordcloud)
library(png)
library(grid)
library(pander)
library(tidyverse)
library(kableExtra)
options(stringsAsFactors = FALSE)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

\newpage

# Introduction

There has been interest in the relationship between second-strand synthesis of AAV genomes and the DNA repair pathways non-homologous end joining (NHEJ) and homology-directed repair (HDR). These pathways may also play a role in the integrity of ITR-genome junctions in integrated vectors.

This preliminary study aims to understand the relationship between HDR and AAV integration integrity. U2OS cells were treated with DMSO or HDR inhibitor B02 (RAD51) prior to AAV transduction. The contribution of purification of the AAV preparations was also explored.

Samples are classified as follows:\newline DMSOCrude: Cells treated with DMSO and transduced with a crude AAV prep \newline DMSOPure: Cells treated with DMSO and transduced with a pure AAV prep\newline HDRiCrude: Cells treated with B02 and transduced with a crude AAV prep\newline HDRiPure: Cells treated with B02 and transduced with a pure AAV prep\newline Negative: Negative Control

Samples were also run with a positive control containing 6 synthetic yeast integration sites, each with breakpoints at the various regions of the ITR dumbbell (DA, DAB, DABB', DABB'C', DABB'C'C, DABB'C'CA)

# Sequencing and Integration Summary Table

Presented below are summary details of each sample. The Unique Molecular Identifiers (UMIs) estimate the number of unique molecules queried. Unique Sites refers to the number of unique integration sites detected. Chao1 estimates the total population sizes of integration sites in the sample.

Prelimary work with negative controls support the use of a filtering parameter in such that sites only supported by one read are considered artifactual and removed.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
knitr::kable(chaoTable, "simple")
```

\newpage

````{=html}
<!--
# Chromosome Map

Locations of integration sites are depicted below by experimental group.

```{r, echo=FALSE, message=FALSE, warning=FALSE}

# create a color key for the plot
group.colors <- c(HDRiCrude = "yellow", DMSOCrude = "red", Negative = "blue", HDRiPure = "green", "DMSOPure" = "orange")

  plot <- ggplot(data = chrom_sizes) + 
    theme(text = element_text(size = 10)) +
    theme(text = element_text(face = "bold")) +
    theme(axis.title = element_text(size = 10)) +
    theme(axis.text = element_text(size = 10)) +
    theme(text=element_text(family="Helvetica")) +
    # base rectangles for the chroms, with numeric value for each chrom on the x-axis
    geom_rect(aes(xmin = as.numeric(chromosome) - 0.2, 
                  xmax = as.numeric(chromosome) + 0.2, 
                  ymax = size, ymin = 0), 
              colour="black", fill = "white", linewidth = 1.0) + 
    # rotate the plot 90 degrees
    coord_flip() +
    # black & white color theme 
    theme(axis.text.x = element_text(colour = "black"), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank()) + 
    # give the appearance of a discrete axis with chrom labels
    scale_x_discrete(name = "chromosome", limits = names(chrom_key)) +
    # add bands for centromeres
    geom_rect(data = centromeres, aes(xmin = as.numeric(chromosome) - 0.2, 
                                      xmax = as.numeric(chromosome) + 0.2, 
                                      ymax = end, ymin = start),colour="black", fill = "black",) +
    # add bands for integration sites
    geom_rect(data = sites, aes(xmin = as.numeric(chromosome) - 0.2, 
                                xmax = as.numeric(chromosome) + 0.2, 
                                ymax = start + 100000, ymin = start, fill = Drug)) + 
    scale_fill_manual(values = group.colors) +
    ggtitle("Site Positions in hg38") +
    theme(plot.title = element_text(hjust = 0.5)) +
    # supress scientific notation on the y-axis
    scale_y_continuous(labels = comma) +
    ylab("region (bp)") +
    facet_wrap_paginate(vars(Drug),
                        nrow = 1,
                        ncol = 1,
                        page = 1)
  print(plot)
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
  # create a color key for the plot
group.colors <- c(HDRiCrude = "yellow", DMSOCrude = "red", Negative = "blue", HDRiPure = "green", "DMSOPure" = "orange")

plot <- ggplot(data = chrom_sizes) + 
    theme(text = element_text(size = 10)) +
    theme(text = element_text(face = "bold")) +
    theme(axis.title = element_text(size = 10)) +
    theme(axis.text = element_text(size = 10)) +
    theme(text=element_text(family="Helvetica")) +
    # base rectangles for the chroms, with numeric value for each chrom on the x-axis
    geom_rect(aes(xmin = as.numeric(chromosome) - 0.2, 
                  xmax = as.numeric(chromosome) + 0.2, 
                  ymax = size, ymin = 0), 
              colour="black", fill = "white", linewidth = 1.0) + 
    # rotate the plot 90 degrees
    coord_flip() +
    # black & white color theme 
    theme(axis.text.x = element_text(colour = "black"), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank()) + 
    # give the appearance of a discrete axis with chrom labels
    scale_x_discrete(name = "chromosome", limits = names(chrom_key)) +
    # add bands for centromeres
    geom_rect(data = centromeres, aes(xmin = as.numeric(chromosome) - 0.2, 
                                      xmax = as.numeric(chromosome) + 0.2, 
                                      ymax = end, ymin = start),colour="black", fill = "black",) +
    # add bands for integration sites
    geom_rect(data = sites, aes(xmin = as.numeric(chromosome) - 0.2, 
                                xmax = as.numeric(chromosome) + 0.2, 
                                ymax = start + 100000, ymin = start, fill = Drug)) + 
    scale_fill_manual(values = group.colors) +
    ggtitle("Site Positions in hg38") +
    theme(plot.title = element_text(hjust = 0.5)) +
    # supress scientific notation on the y-axis
    scale_y_continuous(labels = comma) +
    ylab("region (bp)") +
    facet_wrap_paginate(vars(Drug),
                        nrow = 1,
                        ncol = 1,
                        page = 2)
  print(plot)

```

\newpage

```{r, echo=FALSE, message=FALSE, warning=FALSE}

# create a color key for the plot
group.colors <- c(HDRiCrude = "yellow", DMSOCrude = "red", Negative = "blue", HDRiPure = "green", "DMSOPure" = "orange")

  plot <- ggplot(data = chrom_sizes) + 
    theme(text = element_text(size = 10)) +
    theme(text = element_text(face = "bold")) +
    theme(axis.title = element_text(size = 10)) +
    theme(axis.text = element_text(size = 10)) +
    theme(text=element_text(family="Helvetica")) +
    # base rectangles for the chroms, with numeric value for each chrom on the x-axis
    geom_rect(aes(xmin = as.numeric(chromosome) - 0.2, 
                  xmax = as.numeric(chromosome) + 0.2, 
                  ymax = size, ymin = 0), 
              colour="black", fill = "white", linewidth = 1.0) + 
    # rotate the plot 90 degrees
    coord_flip() +
    # black & white color theme 
    theme(axis.text.x = element_text(colour = "black"), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank()) + 
    # give the appearance of a discrete axis with chrom labels
    scale_x_discrete(name = "chromosome", limits = names(chrom_key)) +
    # add bands for centromeres
    geom_rect(data = centromeres, aes(xmin = as.numeric(chromosome) - 0.2, 
                                      xmax = as.numeric(chromosome) + 0.2, 
                                      ymax = end, ymin = start),colour="black", fill = "black",) +
    # add bands for integration sites
    geom_rect(data = sites, aes(xmin = as.numeric(chromosome) - 0.2, 
                                xmax = as.numeric(chromosome) + 0.2, 
                                ymax = start + 100000, ymin = start, fill = Drug)) + 
    scale_fill_manual(values = group.colors) +
    ggtitle("Site Positions in hg38") +
    theme(plot.title = element_text(hjust = 0.5)) +
    # supress scientific notation on the y-axis
    scale_y_continuous(labels = comma) +
    ylab("region (bp)") +
    facet_wrap_paginate(vars(Drug),
                        nrow = 1,
                        ncol = 1,
                        page = 3)
  print(plot)
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
  
# create a color key for the plot
group.colors <- c(HDRiCrude = "yellow", DMSOCrude = "red", Negative = "blue", HDRiPure = "green", "DMSOPure" = "orange")

plot <- ggplot(data = chrom_sizes) + 
    theme(text = element_text(size = 10)) +
    theme(text = element_text(face = "bold")) +
    theme(axis.title = element_text(size = 10)) +
    theme(axis.text = element_text(size = 10)) +
    theme(text=element_text(family="Helvetica")) +
    # base rectangles for the chroms, with numeric value for each chrom on the x-axis
    geom_rect(aes(xmin = as.numeric(chromosome) - 0.2, 
                  xmax = as.numeric(chromosome) + 0.2, 
                  ymax = size, ymin = 0), 
              colour="black", fill = "white", linewidth = 1.0) + 
    # rotate the plot 90 degrees
    coord_flip() +
    # black & white color theme 
    theme(axis.text.x = element_text(colour = "black"), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank()) + 
    # give the appearance of a discrete axis with chrom labels
    scale_x_discrete(name = "chromosome", limits = names(chrom_key)) +
    # add bands for centromeres
    geom_rect(data = centromeres, aes(xmin = as.numeric(chromosome) - 0.2, 
                                      xmax = as.numeric(chromosome) + 0.2, 
                                      ymax = end, ymin = start),colour="black", fill = "black",) +
    # add bands for integration sites
    geom_rect(data = sites, aes(xmin = as.numeric(chromosome) - 0.2, 
                                xmax = as.numeric(chromosome) + 0.2, 
                                ymax = start + 100000, ymin = start, fill = Drug)) + 
    scale_fill_manual(values = group.colors) +
    ggtitle("Site Positions in hg38") +
    theme(plot.title = element_text(hjust = 0.5)) +
    # supress scientific notation on the y-axis
    scale_y_continuous(labels = comma) +
    ylab("region (bp)") +
    facet_wrap_paginate(vars(Drug),
                        nrow = 1,
                        ncol = 1,
                        page = 4)
  print(plot)

```

\newpage

```{r, echo=FALSE, message=FALSE, warning=FALSE}
# create a color key for the plot
group.colors <- c(HDRiCrude = "yellow", DMSOCrude = "red", Negative = "blue", HDRiPure = "green", "DMSOPure" = "orange")


  plot <- ggplot(data = chrom_sizes) + 
    theme(text = element_text(size = 10)) +
    theme(text = element_text(face = "bold")) +
    theme(axis.title = element_text(size = 10)) +
    theme(axis.text = element_text(size = 10)) +
    theme(text=element_text(family="Helvetica")) +
    # base rectangles for the chroms, with numeric value for each chrom on the x-axis
    geom_rect(aes(xmin = as.numeric(chromosome) - 0.2, 
                  xmax = as.numeric(chromosome) + 0.2, 
                  ymax = size, ymin = 0), 
              colour="black", fill = "white", linewidth = 1.0) + 
    # rotate the plot 90 degrees
    coord_flip() +
    # black & white color theme 
    theme(axis.text.x = element_text(colour = "black"), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank()) + 
    # give the appearance of a discrete axis with chrom labels
    scale_x_discrete(name = "chromosome", limits = names(chrom_key)) +
    # add bands for centromeres
    geom_rect(data = centromeres, aes(xmin = as.numeric(chromosome) - 0.2, 
                                      xmax = as.numeric(chromosome) + 0.2, 
                                      ymax = end, ymin = start),colour="black", fill = "black",) +
    # add bands for integration sites
    geom_rect(data = sites, aes(xmin = as.numeric(chromosome) - 0.2, 
                                xmax = as.numeric(chromosome) + 0.2, 
                                ymax = start + 100000, ymin = start, fill = Drug)) + 
    scale_fill_manual(values = group.colors) +
    ggtitle("Site Positions in hg38") +
    theme(plot.title = element_text(hjust = 0.5)) +
    # supress scientific notation on the y-axis
    scale_y_continuous(labels = comma) +
    ylab("region (bp)") +
    facet_wrap_paginate(vars(Drug),
                        nrow = 1,
                        ncol = 1,
                        page = 5)
  print(plot)
```
-->
````

\newpage

# AAV ITR Breakpoint Plots

These plots are a visual representation of the ITR integration breakpoint. The x-axis indicates the point of departure from the expected ITR, where zero refers to the annotated start of the D-region. The arrow indicates the location within the ITR of the sequencing primer and the dotted lines indicate the tips of the B and C dumbells. Colors indicate the number of rearrangements in the ITR remnant.

DMSOCrude

```{r echo=FALSE, out.width = "50%"}

knitr::include_graphics("Data/pRep~pGTSP5787~GTSP5787.png")
knitr::include_graphics("Data/pRep~pGTSP5792~GTSP5792.png")

```

DMSOPure

```{r echo=FALSE, out.width = "50%"}
knitr::include_graphics("Data/pRep~pGTSP5790~GTSP5790.png")
knitr::include_graphics("Data/pRep~pGTSP5795~GTSP5795.png")
```

\newpage

HDRiCrude

```{r echo=FALSE, out.width = "50%"}
knitr::include_graphics("Data/pRep~pGTSP5786~GTSP5786.png")
knitr::include_graphics("Data/pRep~pGTSP5791~GTSP5791.png")
```

HDRiPure

```{r echo=FALSE, out.width = "50%"}
knitr::include_graphics("Data/pRep~pGTSP5789~GTSP5789.png")
knitr::include_graphics("Data/pRep~pGTSP5794~GTSP5794.png")
```

Negative

```{r echo=FALSE, out.width = "50%"}
knitr::include_graphics("Data/pRep~pGTSP5788~GTSP5788.png")
knitr::include_graphics("Data/pRep~pGTSP5793~GTSP5793.png")
```

\newpage

# Rearrangements in AAV integration sites

This output takes all the sequences of integration sites and evaluates the region between the ITR primer and the host genome junction for alignment to the vector. Any deviation from the expected ITR sequence is considered a rearrangement.   

The plot below depicts the percentage of integration sites with an ITR remnant containing at least one rearrangement.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
knitr::include_graphics("Output/boolean_graph.png")
```

\newpage

# Rearrangements in all ITR containing sequences

This output takes all the ITR containing reads and evaluates them for alignment to the vector. Any deviation from the expected ITR sequence is considered a rearrangement. 

Duplicate reads with the same unique molecular identifiers are collapsed into one molecule for this analysis. The plot below depicts the percentage of unique molecules with an ITR remnant containing at least one rearrangement.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
knitr::include_graphics("Output/anchor_read_graph.png")
```

\newpage

# ITR into ITR vs. ITR into Vector

Evaluating anchor reads that break into vector instead of human genome shows a distribution of of breakpoints into ITR again or into the vector genome. The red lines demarcate the ITR boundaries.

Position 300 on the vector corresponds to the collapsed lacO and mScarlet repeats.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
knitr::include_graphics("Output/distribution.png")
```

\newpage
The relationship between the drug conditions and the ratio of ITR into ITR vs ITR into Vector junctions is evaluated below.

The relationship is significant if the p-value is less than 0.05 and the strength of the relationship is weak at Cramer's V greater than 0.10 and strong at Cramer's V greater than 0.50.

```{r, out.width = "50%", echo=FALSE, message=FALSE,warning=FALSE}
knitr::include_graphics("Output/MosaicCrude.png")
knitr::include_graphics("Output/MosaicPure.png")
knitr::include_graphics("Output/MosaicTotal.png")
knitr::kable(statstable, "simple")
```

\newpage

# Methods

Report Generated on: `r params$date` \newline AAVenger Version: 1.1 \newline \newline Modules Called: 

* core.R  
* mapSiteLeaderSequences.R  
* buildAAVremnantPlots.R  
* anchorReadRearrangements.R


